tests/pfcwd/test_pfcwd_function.py

```diff
@@ -732,6 +732,73 @@ def storm_detect_path(self, dut, port, action):
 
+    def storm_detect_path_cisco_8000(self, dut, port, action):
+        """
+        Storm detection action and associated verifications
+
+        Args:
+            dut(AnsibleHost) : DUT instance
+            port(string) : DUT port
+            action(string) : PTF test action
+
+        Returns:
+            loganalyzer(Loganalyzer) : instance
+        """
+        restore_time = self.timers['pfc_wd_restore_time_large']
+        detect_time = self.timers['pfc_wd_detect_time']
+
+        loganalyzer = LogAnalyzer(ansible_host=self.dut,
+                                  marker_prefix="pfc_function_storm_detect_{}_port_{}".format(action, port))
+        marker = loganalyzer.init()
+        ignore_file = os.path.join(TEMPLATES_DIR, "ignore_pfc_wd_messages")
+        reg_exp = loganalyzer.parse_regexp_file(src=ignore_file)
+        loganalyzer.ignore_regex.extend(reg_exp)
+        loganalyzer.expect_regex = []
+        loganalyzer.match_regex = []
+
+        if action != "dontcare":
+            start_wd_on_ports(dut, port, restore_time, detect_time, action)
+
+        if not self.pfc_wd['fake_storm']:
+            self.storm_hndle.start_storm()
+
+        if action == "dontcare":
+            start_wd_on_ports(dut, port, restore_time, detect_time, "drop")
+
+        # placing this here to cover all action types. for 'dontcare' action,
+        # wd is started much later after the pfc storm is started
+        if self.pfc_wd['fake_storm']:
+            PfcCmd.set_storm_status(dut, self.queue_oid, "enabled")
+        time.sleep(5)
+
+        # Verify that watchdog is not yet triggered.
+        # test pfcwd functionality on a storm, without traffic.
+        if self.pfc_wd['fake_storm']:
+            loganalyzer.expect_regex.extend([EXPECT_PFC_WD_DETECT_RE + fetch_vendor_specific_diagnosis_re(dut)])
+        else:
+            loganalyzer.match_regex.extend([EXPECT_PFC_WD_DETECT_RE + fetch_vendor_specific_diagnosis_re(dut)])
+        loganalyzer.analyze(marker)
+        marker2 = loganalyzer.init()
+        loganalyzer.expect_regex = []
+        loganalyzer.match_regex = []
+        if self.pfc_wd['fake_storm']:
+            loganalyzer.match_regex.extend([EXPECT_PFC_WD_DETECT_RE + fetch_vendor_specific_diagnosis_re(dut)])
+        else:
+            loganalyzer.expect_regex.extend([EXPECT_PFC_WD_DETECT_RE + fetch_vendor_specific_diagnosis_re(dut)])
+
+        # Watchdog was not triggered until this point.
+        # Finally send traffic and trigger watchdog. Until this point,
+        # there was no data traffic. Only pfc storm was being sent.
+        self.traffic_inst.verify_wd_func(action, self.rx_action, self.tx_action)
+        # storm detect
+        logger.info("Verify if PFC storm is detected on port {}".format(port))
+        loganalyzer.analyze(marker2)
+
+        # test pfcwd functionality on a storm after triggering the watchdog.
+        self.traffic_inst.verify_wd_func(action, self.rx_action, self.tx_action)
```

Comment:
Is function "verify_wd_func" used here as background traffic to trigger the pfcwd? 
What are the criteria for background traffic to trigger the pfcwd? Is function "verify_wd_func" good enough here?
In the init function of class SendVerifyTraffic(), it would also call the function "verify_rx_ingress" and send the traffic.

Reply:
@lipxu : No, the pfcwd is triggered as soon as the background traffic and pfc storm are started. If the pfc storm is stopped, the wd should restore. The background traffic will be present through out the entire script.

The background traffic alone cannot trigger the watchdog. Pfc storm is also needed.

verify_wd_func is unchanged - it simply verifies the functionality of pfcwd for traffic - which should be dropped when WD is active, and pass through when WD is inactive.

Reply:
thanks @rraghav-cisco , then which function would take care of the background traffic in this case?

Reply:
@lipxu :  In case of pfcwd_function, the line 798 is where the traffic is actually started. I apologize for confusing, the background traffic is only for the other 2 tests(pfcwd_timer_accuracy and pfcwd_all_port_storm).
In case of pfcwd_function there is no background traffic, the WD is triggered only at line 798, when the traffic is actually sent and WD actually is triggered.

Reply:
So the function "verify_wd_func" here is still used as background traffic to trigger the pfcwd, the function only sends a small number of packages, is it good enough? 
BTW, why is it different from the other two cases? why not using background function here too? or why the other two cases using this function as background traffic?

Reply:
The pfcwd_function test is fundamentally different from the timer_accuracy and all_port_storm. The pfcwd_function verifies that the WD is triggered or not under the right conditions. But the timer_accuracy and all_port_storm *expect* the WD triggered and do their measurements from there.

This PR verifies that the WD is *not* triggered when there is no traffic - So background traffic should not be present. This is the most important requirement for this PR - the PFCWD behaviour has been updated in cisco-8000 to *not* trigger the WD anymore when there is no traffic. That is why this PR was required.

Reply:
1. with background traffic, the PFCWD will be triggered in the event of a storm
2. without background traffic, the PFCWD will not be trigged even in the presence of a storm

There are two conditions here,  as you mentioned, is this test case for the CISCO platform solely to ensure that the PFCWD is not triggered without background traffic? and would not cover the other condition?

Reply:
I think we don't need 2nd here. If we need, it should be in a separate test function,

Reply:
@lipxu : The second condition is only for cisco platforms. The first condition is also partially similar to other platforms, only difference is that we need a traffic to trigger the WD. The other vendors have not modified the pfcwd behaviour AFAIK.

@kevinskwang , a seperate function to cover the second condition is pretty much going to have the first half of the  function. storm_detect_path_cisco_8000(). If this is needed, I will go ahead and do that. Pls let me know.

Reply:
yes, pls just keep the original  test case purpose in this function with 1st condition in this case.